datafusion/core/src/datasource/memory.rs

```diff
@@ -265,6 +265,10 @@ impl TableProvider for MemTable {
         input: Arc<dyn ExecutionPlan>,
         insert_op: InsertOp,
     ) -> Result<Arc<dyn ExecutionPlan>> {
+        if self.batches.is_empty() {
```

Comment:
Since we [allow](https://github.com/apache/datafusion/blob/78ee13a7cfdb741d91be805e4117281a2054d2d3/datafusion/physical-plan/src/execution_plan.rs#L817) scanning a MemTable with zero partitions, it seems appropriate to perform this check here.